src/compiler/nvcc.rs

```diff
@@ -461,7 +461,20 @@ pub fn generate_compile_commands(
         output_file_name: output.file_name().unwrap().to_owned(),
     };
 
-    Ok((command, None, Cacheable::Yes))
+    Ok((
+        command,
+        None,
+        // Never assume the outer `nvcc` call is cacheable. We must decompose the nvcc call into
+        // its constituent subcommands with `--dryrun` and only cache the final build product.
+        //
+        // Always decomposing `nvcc --dryrun` is the only way to ensure caching nvcc invocations
+        // is fully sound, because the `nvcc -E` preprocessor output is not sufficient to detect
+        // all source code changes.
+        //
+        // Specifically, `nvcc -E` always defines __CUDA_ARCH__, which means changes to host-only
+        // code guarded by an `#ifndef __CUDA_ARCH__` will _not_ be captured in `nvcc -E` output.
+        Cacheable::No,
+    ))
```

Comment:
We might now be able to get away with doing less in the `preprocess()` function, since we effectively don't care about most of what it does.